### ISSUE TITLE
Image not showing in study options menu.

### ISSUE THREAD

**Issue by krmanik** [ankidroid/Anki-Android#6669]
###### Reproduction Steps

1. Open AnkiDroid
2. Select select deck 

###### Expected Result
It should show images.

###### Actual Result
It is not showing image.

In Anki Desktop it is showing images. But in AnkiDroid showing small square.
###### AnkiDesktop

![fd](https://user-images.githubusercontent.com/12841290/87274066-5fc1bb00-c50d-11ea-81ae-5a9ed6afd186.PNG)

###### AnkiDroid

![demo](https://user-images.githubusercontent.com/12841290/87273875-d4482a00-c50c-11ea-95d3-b1ef2bdbaab8.png)

###### Debug info
AnkiDroid 2.12beta5

###### Research
*Enter an [x] character to confirm the points below:*

- [ x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [ x ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ x ] I have searched for similar existing issues here and on the user forum
- [ x ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

**Comment by mikehardy** [comment]
For those that aren't sure where this is (it is a screen not seen often in some AnkiDroid configurations) this will show up if you tap the new/due etc numbers on the right side of the deck list screen for a deck, or if you are in "tablet" mode where the main activity goes split screen, at which point it's visible on the right side

**Comment by david-allison-1** [comment]
This isn't currently a WebView, we'll probably need to switch implementations to handle any complex cases.

https://github.com/ankidroid/Anki-Android/blob/017cb41d04e55b284f5e11410c50f63d2b368d38/AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.java#L92

**Comment by infinyte7** [comment]
I will complete this.

**Comment by Ekagra** [comment]
I am working on it

**Comment by Ekagra** [comment]
It's not showing any square on that screen, can anyone help me with that?

**Comment by infinyte7** [comment]
This contain sample deck. Import it in AnkiDroid and click on far right (new, learn, review number) on deck name. 
To view the source. 
1. Long press sample deck name
2. Click option
3. View deck description
In Anki Desktop the description window use qt webview to display html. So, in case of AnkiDroid also android webview will be used.

Extract sample.apkg, import and view
[sample.zip](https://github.com/ankidroid/Anki-Android/files/6128491/sample.zip)


**AnkiDroid**
<img src="https://user-images.githubusercontent.com/12841290/110912366-d0e09900-834e-11eb-9ed3-9c995d9ff748.gif" height=400></img>

**AnkiDesktop**
<img src="https://user-images.githubusercontent.com/12841290/110912409-ddfd8800-834e-11eb-9581-ffdead069618.gif" height=400></img>

**Comment by Ekagra** [comment]
@infinyte7 thanks for the input

**Comment by Ekagra** [comment]
can somebody kindly review this PR @infinyte7 @david-allison-1 ?

**Comment by Abanoub-Asaad** [comment]
I would like to work on this issue.

**Comment by mikehardy** [comment]
@Abanoub-Asaad go for it!

**Comment by Ghanem21** [comment]
is this issue is free now?
what I had understad that I have to replace the square(which is a placeholder) with the image selected with the user

**Comment by krmanik** [comment]
@Ghanem21 
You can create PR for this. You have to check AnkiDesktop also, how it behaves? There are limited HTML tags that work on AnkiDesktop. Earlier webview considered but now it can be implemented without webview.

**Comment by akp660** [comment]
I can help with this by solving the image preview issues.

**Comment by dae** [comment]
IIRC recent desktop versions don't support this either.

**Comment by dae** [comment]
Sorry, I remembered incorrectly. Images are disabled if you have markdown enabled, but that may change in the future.

**Comment by ShadyMedic** [comment]
Thank you for your answer any reopening this.

Please, don't disable this feature in the desktop version as well. It makes no sense, when remote images are allowed on cards and removing the support for them on cards as well (to make the "privacy and tracing" concerns valid) would make a lot of people mad.

There's a way to go around this restriction (for this "update notification" feature) by adding a card to the deck, which will load the remote image with the current update status and then prompt the user to bury it, but it just makes it harder for everyone (plus, let's be honest, most Anki users are causals who don't even know what "Bury card" even means).

**Comment by wyy511511** [comment]
I would like to say, it's an excellent and important feature, and @ShadyMedic has ingeniously utilized it. I definitely need this feature!



# Desktop
Version ⁨2.1.65 (aa9a734f)⁩
Python 3.9.15 Qt 6.5.0 PyQt 6.5.0

## deck's info page
works great on deck's info page
<img width="675" alt="image" src="https://github.com/ankidroid/Anki-Android/assets/45887805/003acf9d-c9b7-41e6-aaa8-125d55133870">

## card view
The image resources are not rendered and displayed in the card view. I haven't used the devtool to check yet.
<img width="1024" alt="img1" src="https://github.com/ankidroid/Anki-Android/assets/45887805/9f3c208b-0734-43ed-b90e-5943c98d72de">


# AnkiDroid 

AnkiDroid Version = 2.18alpha6-debug (a854e93b5f34b0ea167806787375c2d7ab43205c)
Backend Version = 0.1.34-anki23.12.1 (23.12.1 1a1d4d5419c6b57ef3baf99c9d2d9cf85d36ae0a)

## deck's info page

<img width="188" alt="iShot_2024-03-24_04 52 31" src="https://github.com/ankidroid/Anki-Android/assets/45887805/62838f82-9c54-4e9e-befe-9d4f9416fc1b">

<details>
  <summary> How to set up a test deck with url img?</summary>
  
  The UI and sequence shown in the gif below have become deprecated.
  > <img alt="" height="400" src="https://user-images.githubusercontent.com/12841290/110912366-d0e09900-834e-11eb-9ed3-9c995d9ff748.gif"> 
</details>

<img width="840" alt="iShot_2024-03-24_05 07 09" src="https://github.com/ankidroid/Anki-Android/assets/45887805/91bb63f0-54e5-4fd0-8485-f5637a0b886d">




Type **`chrome://inspect`** in your browser 

<img width="794" alt="iShot_2024-03-24_04 46 33" src="https://github.com/ankidroid/Anki-Android/assets/45887805/70487e2a-b62e-44f6-bb9d-6d625ed0b8f4">


<img width="992" alt="iShot_2024-03-24_04 42 40" src="https://github.com/ankidroid/Anki-Android/assets/45887805/a6c7f746-252f-4b17-abf0-a7b83f35089e">


Resource Loading Policy: CORS?

Path Issues: base64?  downloaded locally and then loaded into the WebView from a local path? 

Permission Issues: user need give permissions to folder/resources.

WebView Configuration Issues:  WebView configuration that prohibit image loading.
I guess Anki may have implemented such settings, but I'm not entirely sure.

Image Format,Network Issues etc.     This does not apply to this issue.

**Comment by ShadyMedic** [comment]
@wyy511511 Hi there, thank you for your support.

The reason why you don't see the image in the card on desktop is because Anki parses the card fields on order to display exactly what you typed in (which is desirable most of the time). In order to make the HTML code render an image, you need to paste it into the code editor (you can enable it by clicking the `<>` icon or by pressing `Ctrl+Shift+X`).

About deck description in AnkiDroid, I looked into the code some time ago and although I don't umderstand mobile app development all that much, it seemed to me like the deck description is not actually being rendered by WebView and there is an parser function in the code somewhere, which reformats certain HTML elements to better fit the app and override user-defined CSS (for example, by removing borders and background colors). Remote images are disabled at this step as well, I think.

**Comment by Arthur-Milchior** [comment]
@wyy511511 Thank you very much for the reproduction step.
Do you think you could export this deck and share it (you need to zip it first, otherwise github do not accept the file) so other people can save time here.

**Comment by wyy511511** [comment]
> This isn't currently a WebView, we'll need to switch implementations to handle any complex cases.
> 
> https://github.com/ankidroid/Anki-Android/blob/017cb41d04e55b284f5e11410c50f63d2b368d38/AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.java#L92

Hi @david-allison,

Switching to a WebView can solve this problem, but based on your response, do you prefer solving it by customizing the TextView and using Html.ImageGetter?
If AnkiDroid is okay with switching to WebView, I think the related test code also needs to be modified. 
<details>
  <summary>TextView related test code</summary>

https://github.com/wyy511511/Anki-Android/blob/23b507a2e89b4b6c4ff9cc4f76e0db1f47c08fae/AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt#L690-L698

https://github.com/wyy511511/Anki-Android/blob/bd9b7013fdfff1c1149622599e865ea2f728f0ed/AnkiDroid/src/test/java/com/ichi2/anki/StudyOptionsFragmentTest.kt#L24-L88

</details>


I'm not familiar with the testing part yet, so I'm hesitant to make changes directly. If needed, I'd be happy to work on the testing part.




<img width="327" alt="image" src="https://github.com/ankidroid/Anki-Android/assets/45887805/9deed8e0-0a2c-4119-b5e6-2f61bba8ba19">

**Comment by wyy511511** [comment]
Test desk apkg
[AnkiDroid Issues#6669.apkg.zip](https://github.com/ankidroid/Anki-Android/files/14879347/AnkiDroid.Issues.6669.apkg.zip)

**Comment by david-allison** [comment]
We should aim to follow Anki Desktop's current implementation. 

First investigate what this is, especially the security procedures taken.

This may have changed in the past 4 years 

Regarding tests, feel free to leave `@NeedsTest` in appropriate locations, we're only at around 40% coverage, but we want to improve over time. If you can recreate the "logic" of the tests, then that is encouraged. If they're no longer relevant then they can be removed

**Comment by wyy511511** [comment]
I reviewed [Anki Desktop's implementation](https://github.com/ankitects/anki/blob/429bc9e14cefb597646a0e1beac6ef140f226b6f/qt/aqt/overview.py#L199) , where the original `QPlainTextEdit()` string is wrapped in a <div> tag within the` _desc()` function(this includes a branch to support markdown). The generated HTML string is then integrated into the Overview page's content and displayed in a WebView (i.e., self.web) through the `self.web.stdHtml` method.

I'll update tests.

**Comment by wyy511511** [comment]
> Future versions of Anki will block JS from the deck descriptions - the fact that it's currently allowed was just an oversight.
_Originally posted by @dae in https://github.com/ankidroid/Anki-Android/issues/8184#issuecomment-799007081_


I tested Anki Desktop still allow js.

**Comment by asp616848** [comment]
Is the issue resolved now? I should be able to help with my android development knowledge.

**Comment by ShadyMedic** [comment]
No, the issue still persists. It would be awesome if you could fix it.

7. 8. 2024 7:21:57 Abhijeet Shashwat ***@***.***>:

**Comment by ShadyMedic** [comment]
That doesn't make sense though, since external images are already allowed on the cards and as far as I know, JavaScript works as well.

And I am not aware of any security incidend that this has caused. So if this was allowed in deck descriptions as well (that many users don't even see because they go straight to browsing the cards), it wouldn't worsen the security of the app in any way.

7. 8. 2024 9:34:11 Yaoyao ***@***.***>:

**Comment by shkshreyas** [comment]
please assign me this issue

**Comment by david-allison** [comment]
Before starting, please evaluate how Anki Desktop implements this feature, and post it as a comment

**Comment by ShaanNarendran** [comment]
I did look into this for a while now, I think I have an idea of a potential solution. Basically from what I saw in Anki, when we pass our image as an img into the description, it is processed by aqt/deckdescription.py and then when overview.py goes to display it, the img tag is automatically processed by qt in overview.py and we get the image in the app. Someone on the thread mentioned ImageGetter and i think a possible solution could be to implement the ImageGetter interface and use that when we format the description. This lets us stick to using text view and as far as security goes I'm definitely not experienced enough to have a proper say but the input is already sanitised in both Anki and AnkiDroid so there would be no worries about script injections. If the solution seems viable, I'm willing to try it out if the issue is not assigned to anyone already. If there any clarifications, I'd be happy to explain more since there's already alot of info here. Also, I'm not sure how accurate my analysis is since this is my first time doing an analysis like this, so please feel free to correct me.

**Comment by ShaanNarendran** [comment]
I think I could probably try and fix this, could I be assigned the issue?

**Comment by david-allison** [comment]
Knowing what we know now, I don't think this is a good first issue:

* Markdown descriptions are now supported (Anki 2.41+), and if I recall correctly, images are no longer usable 
* A deck may be loading external resources, and we want to discuss how/whether to support this
  * And whether we can or should put effort into the legacy description code path

Needs more discussion/research on this issue before a solution is coded

**Comment by ShaanNarendran** [comment]
@david-allison  I actually did manage to display and load the sample image and a few other images I found online. By implementing the ImageGetter interface and passing it as a parameter to the htmlcompat in the formatdescription function, you can pretty much load any image which has a http/https, it's still using textview not webview so I believe its fairly safe. I'm not really sure of the processes but could send some more info in the discord or open a pr to show what I did.

<img width="344" height="773" alt="Image" src="https://github.com/user-attachments/assets/37b730c0-362e-43f6-8743-b3e82f484c54" />

**Comment by david-allison** [comment]
Definitely send in a PR, great going! But I can't guarantee it'll be merged.

The problem with HTTPS images is that it would mean we supports [tracking pixels](https://en.wikipedia.org/wiki/Web_beacon#Spy_pixels), which can be used to breach privacy.

**Comment by ShadyMedic** [comment]
I said it before and I'll say it again.

Enabling external images in deck descriptions doesn't make the app any less secure than it already is, because external images, including any tracking ones are already possible on the cards themselves.

Also, what possible dangerous tracking are we subjecting users to? Revealing their IP address? I can do the same by sending them the deck as a link to a redirecting IP grabber. Revealing when they are opening a certain deck? That's very unreliable since most AnkiDroid users open the deck directly and don't even see the description. Fingerprinting? Okay, I guess we could make Anki send the requests with uniform headers so they appear as the same user/browser to the remote servers (besides the IP addressed).

I'm actually very into security and privacy, I'm literally typing this response within FairEmail app, but pragmatically speaking, this feature has nothing to do with decreasing/increasing security.

7. 12. 2025 18:56:28 David Allison ***@***.***>:

**Comment by david-allison** [comment]
* The deck description is displayed alongside the deck picker on larger devices, enabling loading URLs here would mean that the URL would be accessed before the user has a chance to inspect the deck via 'Browse', Manage Note Types, or the Note Editor.
* Editing the deck description is not commonly done by users (aside from deck authors) - most wouldn't know/bother to look there for trackers, or expect that it's possible
* It's unexpected (to me) that the deck description should be contacting an external server. 
  * Why can't the image be included statically in the deck media?
* It's unexpected that the deck description would be displayed differently when offline

**Comment by ShaanNarendran** [comment]
> * The deck description is displayed alongside the deck picker on larger devices, enabling loading URLs here would mean that the URL would be accessed before the user has a chance to inspect the deck via 'Browse', Manage Note Types, or the Note Editor.
> * Editing the deck description is not commonly done by users (aside from deck authors) - most wouldn't know/bother to look there for trackers, or expect that it's possible
> * It's unexpected (to me) that the deck description should be contacting an external server.
>   
>   * Why can't the image be included statically in the deck media?
> * It's unexpected that the deck description would be displayed differently when offline

This is very valid, the fix I implemented only works if online, I didn't really think about the offline part till you just brought it up. I think that would be alot more complex, but I hope my pr can provide some help.

**Comment by ShadyMedic** [comment]
> Editing the deck description is not commonly done by users (aside from deck authors) - most wouldn't know/bother to look there for trackers, or expect that it's possible

Well editing notes in shared decks, that are usually as good as it gets is not common either. If anything, the users will want to mark individual notes and they can already use stars or flags for it, so there's no motivation for them to view and edit card source.

Also, an average user will not see a difference between `<img src="brainstem.png"/>` and `<img src="http://anki.my.net/img?id=6669"/>`. The ones that do are tech-savvy and would likely check the deck description as well.

> Why can't the image be included statically in the deck media?

Loading media from external servers can trigger functionalities that local static media can't (like my update notification feature mentioned above). In other cases, I have honestly no idea why users create decks with remote images on static web servers (the "makes package file smaller" is not really that much of an advantage) and I was honestly surprised that Anki (desktop) allows such linking, but since it already does and we're trying to make AnkiDroid as compatible/similar with desktop version as possible, I think that we should take it as a fact and it's a valid reason to have the same privacy requirements for deck description images as well. A malicious deck creator will just add a tracking pixel to all notes in the package anyway.

> It's unexpected that the deck description would be displayed differently when offline

Same with images within notes. Same argument as the previous paragraph.

**Comment by Akira2206** [comment]
Hi @david-allison, I would like to work on this issue.

I have read the previous discussion and understand the concerns regarding privacy (tracking pixels) and offline consistency. I have built a working prototype that addresses these specific points without adding any external libraries.

For the technical approach, I implemented a custom `Html.ImageGetter` using `HttpURLConnection`.

Here is how it handles the main requirements:
* To protect privacy, it explicitly strips `Cookie` and `Referer` headers from all requests.
* To support offline usage, it implements a hashed disk cache so remote images persist even in Offline Mode.
* It also correctly resolves local filenames against the `collection.media` directory.

**Proof of Concept:**
I have verified this locally. The video below demonstrates the test: loading a remote image, restarting the app without an internet connection, and confirming the cache works.

https://github.com/user-attachments/assets/64f750c3-2d86-43c7-a59e-9cdc4e36ab55

I believe this covers the main requirements, but as I am relatively new to the project, I would really value your guidance on whether this is the right direction before proceeding.

**Comment by ShadyMedic** [comment]
> I have read the previous discussion and understand the concerns regarding privacy (tracking pixels) and offline consistency. I have built a working prototype that addresses these specific points without adding any external libraries.
> 
> For the technical approach, I implemented a custom `Html.ImageGetter` using `HttpURLConnection`.
> 
> Here is how it handles the main requirements:
> 
>     * To protect privacy, it explicitly strips `Cookie` and `Referer` headers from all requests.
> 
>     * To support offline usage, it implements a hashed disk cache so remote images persist even in Offline Mode.

Amazing work, you're a hero 👏. I have just a few more questions/ideas. You're very likely more experienced in terms of Android app development then I am, so it's likely some of these cannot be answered, sorry about that if that's the case.

1. When the remote image is cached and the device is online, will the client try to load the image from the server again, or will it keep using the cache? If the latter, what's the lifetime of the cache?
2. What is the User-Agent of the requests for the images? I am not sure what do these HTTP libraries set into User-Agent field, but if it contains things like Android version, device codename etc, it could be very useful for fingerprinting (which is a thread that needs to be mitigated). On the other hand, I could see some potential uses if the user agent contained the app version, something like `User-Agent: AnkiDroid-v2.23.2`, but that's not really necessary I guess.

Edit: Ad 1: thinking about it, if the cache is not being reused when the device is online, I'd be in favor of making the cache usable for 1 day (or infinite as long as the device is offline). I can imagine people viewing the same card 20 times in a single study session and if the image is 2 MB, it can really drain ones mobile data.

**Comment by david-allison** [comment]
I don't support performing HTTP[S] requests in the deck description, primarily due to privacy concerns. See also: [my other concerns](https://github.com/ankidroid/Anki-Android/issues/6669#issuecomment-3623132000).

Local images included in the deck's `collection.media` should be sufficient.

Pinging @ankidroid/core

**Comment by ShadyMedic** [comment]
> I don't support performing HTTP[S] requests in the deck description, primarily due to privacy concerns.

With all due respect, these privacy concerns make little sense. A deck description is like... the worst place to place tracking pixels, since it's not even loaded most of the time.

> It's unexpected that the deck description would be displayed differently when offline

Akira2206 resolved this gracefully with the caching feature.

Your other points were countered [here](https://github.com/ankidroid/Anki-Android/issues/6669#issuecomment-3623175049).

If I understood it correctly, you're trying to bring AnkiDroid as close to desktop version of Anki as possible, yet you're refuse to do so in this specific case.

Seeing the number of people in this thread who would like to have this feature, going as far as implementing it themselves, I really don't see any reason to keep blocking it.

**Comment by Akira2206** [comment]
@ShadyMedic Thanks for the support! To answer your questions:

> 1. When the remote image is cached and the device is online, will the client try to load the image from the server again, or will it keep using the cache? If the latter, what's the lifetime of the cache?
>
> Edit: Ad 1: thinking about it, if the cache is not being reused when the device is online, I'd be in favor of making the cache usable for 1 day (or infinite as long as the device is offline). I can imagine people viewing the same card 20 times in a single study session and if the image is 2 MB, it can really drain ones mobile data.

The implementation checks the disk cache first, regardless of connection status. If the image has been downloaded once, it will always load from the local cache on subsequent views. It will not hit the server again for that URL unless the cache is cleared. This minimizes data usage and mitigates "tracking pixels" firing on every view.

The cache uses Android's standard `context.cacheDir`. The files persist indefinitely until the Android OS reclaims space (Low Storage triggers) or the user manually clears the app cache.

> 2. What is the User-Agent of the requests for the images? I am not sure what do these HTTP libraries set into User-Agent field, but if it contains things like Android version, device codename etc, it could be very useful for fingerprinting (which is a thread that needs to be mitigated). On the other hand, I could see some potential uses if the user agent contained the app version, something like `User-Agent: AnkiDroid-v2.23.2`, but that's not really necessary I guess.

To specifically address this, I hardcoded the header to `User-Agent: AnkiDroid`. It does not include the Android version, device model, or app version, making it uniform for all users.

@david-allison I completely understand the privacy stance. While direct HTTP requests cannot hide the user's IP address, this implementation strips Cookies and Referers, and enforces a generic User-Agent, making it significantly more private than a standard WebView or browser.

Just to clarify, this implementation also fixes the resolution of Local Images.
Currently, the codebase struggles to load local images in the description because the standard `Html.fromHtml` doesn't know to look inside the `collection.media` directory. My implementation explicitly resolves filenames against that directory.
If the decision is made to strictly ban remote images, I can easily modify this implementation to only include the Local Image fix (and the background loading to prevent UI lag), while disabling the HTTP path entirely.

**Comment by ShadyMedic** [comment]
Thank you for your answers @Akira2206.

Seems like sensible defaults, my only concern is the infinite lifetime of the cache. I'd be in favor of limiting it to 1 day to 1 week.

One thing is that this "request once only" feature would break my little update subscribe service, but a bigger problem might be caused to decks that utilize these remote images in a more complex way. I think that Amboss is going as far as keeping login token to their service in the cards and serves content from their servers right into the cards based on information about the user (or something like that, I didn't really look to deep into that, I think they even have their own extension).

As I said, there is no reason for people to include static images in the deck by remote-linking them. The only reason to link an image from a remote service is if you want to use it for a dynamic feature like mentioned above, which the infinite caching breaks.

Though, I must admit that I don't know how the desktop version of Anki is handling caching. Whether it loads the remote media again every time, honors the `cache-control: max-age=...` header or uses the cache indefinitely`.

**Comment by Akira2206** [comment]
I updated my implementation to only load local images inside the deck description.
I hadn't considered the OutOfMemory exception that can occur if high-resolution images are used, so I added bitmap downsampling and tested it.
Without the fix, loading ~40 images caused the process to crash. I could load more images with the fix, however it triggered an ANR. I have attached logs that indicates main thread was locked for 5 seconds.

<details>


```
2025-12-21 05:48:54.675   597-597   lowmemorykiller         lmkd                                 E  [ 9500] 10212   4080916    946516              0 com.ichi2.anki.debug

---------------------------- PROCESS STARTED (16584) for package com.ichi2.anki.debug ----------------------------

---------------------------- PROCESS ENDED (9500) for package com.ichi2.anki.debug ----------------------------

---------------------------- PROCESS ENDED (16584) for package com.ichi2.anki.debug ----------------------------

2025-12-21 05:49:02.096  1948-16198 ActivityManager         system_server                        E  ANR in com.ichi2.anki.debug (com.ichi2.anki.debug/com.ichi2.anki.StudyOptionsActivity)
                                                                                                    PID: 9500
                                                                                                    Reason: Input dispatching timed out (8a0d802 com.ichi2.anki.debug/com.ichi2.anki.StudyOptionsActivity (server) is not responding. Waited 5005ms for FocusEvent(hasFocus=true)).
                                                                                                    Parent: com.ichi2.anki.debug/com.ichi2.anki.StudyOptionsActivity
                                                                                                    ErrorId: 139943bd-081c-4ec9-9151-1dddda62d477
                                                                                                    Frozen: false
                                                                                                    Load: 73.61 / 39.97 / 29.72
```

</details>


The implementation is stable for normal usage, but the anti-abuse issue remains.

**Comment by ShaanNarendran** [comment]
> I updated my implementation to only load local images inside the deck description. I hadn't considered the OutOfMemory exception that can occur if high-resolution images are used, so I added bitmap downsampling and tested it. Without the fix, loading ~40 images caused the process to crash. I could load more images with the fix, however it triggered an ANR. I have attached logs that indicates main thread was locked for 5 seconds.
> 
> Details
> The implementation is stable for normal usage, but the anti-abuse issue remains.

Have you made a pr? I think it's best to make one so that your code can be reviewed.

**Comment by Akira2206** [comment]
I'll open one later today. I still want to try a few changes on my end first. I just posted the findings above to keep a record as discussed on Discord.
